### PR TITLE
Store batches from rollups. Check batches at same height are consistent.

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

- Provide a description and a link to the underlying ticket

### What changes were made as part of this PR:

- Provide a high level list of the changes made

### Definition of done

- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated 
- [ ] End-to-end tests run if required (see below)

### End-to-end tests

Running the end-to-end tests in the [obscuro-test repo](https://github.com/obscuronet/obscuro-test) is currently at the 
discretion of the developer prior to merging a PR. The intention is not to slow down development by having the longer 
running end-to-end tests as a PR gate run on each branch commit etc. To manually trigger a run pre-merge;

- Go to [run_local_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_local_tests.yml)
- Click the "Run workflow" button
- Use the main branch of obscuro-test 
- Enter the name of the PR branch for go-obscuro
- Run the workflow

The run takes approximately 20 minutes and any failure will be reported in the workflow output. Should the tests fail 
the docker container output for the local testnet and all test artifacts will be added to the run and can be downloaded
for debugging and analysis, with a retention period of 2 days. 

Note that every PR merge to main will also trigger a run of post-merge tests, so even if you do not manually trigger 
pre-merge, tests will be run post-merge. The intention being to catch any issues fast on main to allow for quick 
resolution. The post-merge test output can be seen at 
[run_merge_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_merge_tests.yml) and will show both 
the PR number and author in the list of runs.

### CODE REVIEW

File: go/enclave/rollupchain/rollup_chain.go [hunk 1]
```diff
@@ -909,6 +886,47 @@ func (rc *RollupChain) checkRollupsCorrectlyChained(rollup *core.Rollup, previou
...
+	//  contains the same transactions, regardless of reorgs), uncomment this check.
+	//if err == nil && batch.Header.TxHash != storedBatch.Header.TxHash {
+	//	return fmt.Errorf("two batches at same height did not have the same transactions")
+	//}
+
+	// We store the batch and update the head batch for that L1 block.
+	if err = rc.storage.StoreBatch(batch, txReceipts); err != nil {
+		return fmt.Errorf("failed to store batch. Cause: %w", err)
+	}
+	if err = rc.storage.UpdateHeadBatch(batch.Header.L1Proof, batch, txReceipts); err != nil {
```
Comment (BedrockSquirrel) [inline go/enclave/rollupchain/rollup_chain.go:923]: Wouldn't this rollback the head of the L2 chain if it was processing a historical rollup batch?

File: go/enclave/rollupchain/rollup_chain.go [hunk 2]
```diff
@@ -909,6 +886,47 @@ func (rc *RollupChain) checkRollupsCorrectlyChained(rollup *core.Rollup, previou
...
+		return fmt.Errorf("could not fetch batch. Cause: %w", err)
+	}
+	// TODO - #718 - Once the sequencer includes transactions deterministically (i.e. a batch of a given height always
+	//  contains the same transactions, regardless of reorgs), uncomment this check.
+	//if err == nil && batch.Header.TxHash != storedBatch.Header.TxHash {
+	//	return fmt.Errorf("two batches at same height did not have the same transactions")
+	//}
+
+	// We store the batch and update the head batch for that L1 block.
+	if err = rc.storage.StoreBatch(batch, txReceipts); err != nil {
```
Comment (BedrockSquirrel) [inline go/enclave/rollupchain/rollup_chain.go:920]: Is it worth checking that we've not already seen the exact same batch before storing it? We'll end up storing every batch twice (no-op but don't know if it's free with our db setup)

**Review by BedrockSquirrel — COMMENTED** [review]
Couple thoughts

**Review by BedrockSquirrel — APPROVED** [review]
LGTM I think, can't wrap my head around whether there's an edgecase where the chain could get stuck because we've seen a block before but it was no longer canonical so may be worth adding that if statement unless I'm wrong.

File: go/enclave/l2chain/l2_chain.go [hunk 3]
```diff
@@ -909,6 +886,51 @@ func (rc *RollupChain) checkRollupsCorrectlyChained(rollup *core.Rollup, previou
...
+	//}
+
+	// If we haven't stored the batch before, we store it and update the head batch for that L1 block.
+	// TODO - FetchBatch should return errutil.ErrNotFound for unstored batches, so we can handle that type of error
+	//  separately.
+	if _, err = lc.storage.FetchBatch(*batch.Hash()); err != nil {
+		if err = lc.storage.StoreBatch(batch, txReceipts); err != nil {
+			return fmt.Errorf("failed to store batch. Cause: %w", err)
+		}
+		if err = lc.storage.UpdateHeadBatch(batch.Header.L1Proof, batch, txReceipts); err != nil {
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/l2chain/l2_chain.go:926]: I don't think there's a right way to do it, unless you're supposed to look at the timestamp or something. Sometimes you'll be processing a newer one, and sometimes an outdated one. I'm going to merge this and then hopefully soon it becomes irrelevant based on your work?

**Pull request merged**